aws-auth-cognito/src/main/java/com/amplifyframework/auth/cognito/asf/UserContextDataProvider.kt

```diff
@@ -0,0 +1,92 @@
+import android.util.Log
+import com.amplifyframework.statemachine.codegen.data.UserPoolConfiguration
+import org.json.JSONException
+import org.json.JSONObject
+
+/**
+ * Provides the user context data that is sent to the server.
+ * @param context android application context
+ * @param configuration config containing cognito userPoolId for the application
+ * and secret key (for now, this would be application clientId) used while generating signature.
+ */
+class UserContextDataProvider(private val context: Context, private val configuration: UserPoolConfiguration) {
+    companion object {
+        private val TAG = UserContextDataProvider::class.java.simpleName
+        private const val VERSION_VALUE = "ANDROID20171114"
+
+        private const val CONTEXT_DATA = "contextData"
+        private const val USERNAME = "username"
+        private const val USER_POOL_ID = "userPoolId"
+        private const val TIMESTAMP_MILLI_SEC = "timestamp"
+        private const val DATA_PAYLOAD = "payload"
+        private const val VERSION_KEY = "version"
+        private const val SIGNATURE = "signature"
+    }
+
+    private val timestamp = System.currentTimeMillis().toString()
+
+    private val aggregator = ContextDataAggregator
+
+    @Throws(JSONException::class)
+    private fun getJsonPayload(contextData: Map<String, String?>, username: String, userPoolId: String): JSONObject {
+        val payload = JSONObject()
+        payload.put(CONTEXT_DATA, JSONObject(contextData))
+        payload.put(USERNAME, username)
+        payload.put(USER_POOL_ID, userPoolId)
+        payload.put(TIMESTAMP_MILLI_SEC, timestamp)
+        return payload
+    }
+
+    @Throws(JSONException::class)
+    private fun getJsonResponse(payload: String, signature: String): JSONObject {
+        val jsonResponse = JSONObject()
+        jsonResponse.put(DATA_PAYLOAD, payload)
+        jsonResponse.put(SIGNATURE, signature)
+        jsonResponse.put(VERSION_KEY, VERSION_VALUE)
+        return jsonResponse
+    }
+
+    private fun getEncodedResponse(jsonResponse: JSONObject): String {
+        val bytes = jsonResponse.toString().toByteArray()
+        return Base64.encodeToString(bytes, Base64.DEFAULT)
+    }
+
+    /**
+     * Gets aggregated user context data, adds signature to it and provides
+     * it in Base64 encoded form. Final data is JSON object with 'signature' and
+     * 'payload'. Payload is a JSON object that contains 'username',
+     * 'userPoolId', 'timestamp' and 'contextData'.
+     * @param username username for the user
+     * @return base64 encoded userContextData.
+     */
+    fun getEncodedContextData(username: String) = try {
+        val contextData = aggregator.getAggregatedData(context)
+        val payload = getJsonPayload(contextData, username, configuration.poolId!!)
+        val payloadString = payload.toString()
+        val signature = SignatureGenerator.getSignature(payloadString, configuration.appClient!!, VERSION_VALUE)
```

Comment:
Is there a reason why we are using `!!` here?

Reply:
`appClient` is required for all authentication APIs, so ideally Auth plugin would have failed at configuration before it could reach here.